src/io/ctrees_utils.c

```diff
@@ -51,17 +49,17 @@ int64_t read_forests(const char *filename, int64_t **f, int64_t **t)
             continue;
         } else {
             const int nitems_expected = 2;
-            XASSERT(ntrees_found < ntrees, -1,
+            XRETURN(ntrees_found < ntrees, -1,
```

Comment:
Also, return value should be `EXIT_FAILURE`.

Reply:
Has to be a negative value, otherwise the caller will not know that there was an error. I can make the code return `-EXIT_FAILURE` - would that work?

Reply:
Ahhh ok; is it a case that this function is returning an actual value rather than simply performing a task and then returning a status? 

In this scenario, would there be any advantage to passing the value to update as a pointer, updating it inside the function, and then returning an explicit exit status?

My question is basically: Is there a guideline/convention for returning an actual value versus passing a pointer and updating in the function itself?

Reply:
I think that is a personal preference -- you could always use the pattern that the return value is to indicate success/failure (or some finer status indication), while any actual needed values are passed via pointers. This works, as long as the developers are comfortable with pointers/pointers-to-pointers/ etc